avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -188,14 +200,20 @@ protected Object deserialize(boolean includeSchemaAndVersion, String topic, Bool
    * @throws SerializationException
    */
   protected GenericContainer deserializeWithSchemaAndVersion(String topic, boolean isKey, byte[] payload) throws SerializationException {
-    return (GenericContainer) deserialize(true, topic, isKey, payload);
+    return (GenericContainer) deserialize(true, topic, isKey, payload, null);
   }
 
-  private DatumReader getDatumReader(Schema writerSchema) {
+  private DatumReader getDatumReader(Schema writerSchema, Schema readerSchema) {
     if (useSpecificAvroReader) {
-      return new SpecificDatumReader(writerSchema, getReaderSchema(writerSchema));
+      if (readerSchema == null) {
+        readerSchema = getReaderSchema(writerSchema);
```

Comment:
Given my other comment about how this is used, I think it probably makes sense if the reader schema is fixed for a deserializer. In that case, we shouldn't resolve `readerSchema` with ever message we deserialize since it won't change over the lifetime of the deserializer.